File: server/lib/lint-passive.py
#!/usr/bin/env python3
"""
Passive-voice detector for .tex files using spaCy dependency parsing.

Reads a TeX file, strips math environments and inline math, runs spaCy on
each prose line, and reports passive-voice findings via dependency parsing
(auxpass / nsubjpass relations).

Usage:
    python3 lint-passive.py <tex-file>

Outputs JSON to stdout: a list of {file, line, snippet, kind} objects.
"""

import json
import re
import sys

import spacy

# Math environment names — content between \begin{...} ... \end{...} is skipped.
ENV_BLOCK = re.compile(
    r"^\s*\\(?:begin|end)\{(?:equation\*?|align\*?|gather\*?|multline\*?|"
    r"displaymath|eqnarray\*?|array)\}"
)

INLINE_MATH_PATTERNS = [
    re.compile(r"\$\$[^$]*\$\$"),
    re.compile(r"\\\[[^\]]*\\\]"),
    re.compile(r"\$[^$\n]*\$"),
    re.compile(r"\\\([^)]*\\\)"),
]

CITE_PATTERNS = [
    re.compile(r"\\(cite[ptes]?|ref|eqref|pageref|cref|Cref|autoref|label)\{[^}]*\}"),
]


def strip_math_and_refs(line: str) -> str:
    """Replace inline math with placeholder 'M', drop cite/ref args."""
    s = line
    for pat in INLINE_MATH_PATTERNS:
        s = pat.sub(" M ", s)
    for pat in CITE_PATTERNS:
        s = pat.sub(" ", s)
    # Strip simple emph/textbf/textit wrappers but keep their text
    s = re.sub(r"\\(emph|textbf|textit|texttt|mathrm|mathit)\{([^}]*)\}", r" \2 ", s)
    return s


# Verbs that are commonly used as state-of-being predicates in math prose
# when no explicit agent is given. "X is bounded" / "X is satisfied" /
# "X is well-defined" usually mean the property holds, not that someone
# performed an action. We only flag these when there's an explicit agent
# ("by Y"), where the agentive reading is unambiguous.
STATIVE_OK_VERBS = {
    "bounded", "satisfied", "defined", "well-defined", "specified",
    "known", "fixed", "named", "labeled", "labelled", "called",
    "needed", "required", "guaranteed", "ensured", "established",
    "true", "false", "unique", "finite", "infinite", "non-empty",
}


def has_agent(verb_token):
    """Return True if the verb has an explicit `by AGENT` child (agentive passive)."""
    for child in verb_token.children:
        if child.dep_ == "agent":
            return True
        # spaCy sometimes uses "prep" with text "by" rather than "agent"
        if child.dep_ == "prep" and child.text.lower() == "by":
            return True
    return False


def lint_text(text: str, file_label: str = "<text>"):
    """Run passive-voice detection line by line."""
    nlp = spacy.load("en_core_web_sm")
    findings = []
    in_math_env = False
    for i, line in enumerate(text.split("\n"), start=1):
        if not in_math_env and ENV_BLOCK.search(line) and "\\begin" in line:
            in_math_env = True
            continue
        if in_math_env:
            if "\\end" in line and ENV_BLOCK.search(line):
                in_math_env = False
            continue
        stripped = line.strip()
        if not stripped or stripped.startswith("%"):
            continue
        prose = strip_math_and_refs(line)
        if not prose.strip():
            continue
        doc = nlp(prose)
        # Iterate tokens; flag passive auxiliaries with their verb.
        seen_verbs = set()
        for tok in doc:
            if tok.dep_ == "auxpass":
                verb = tok.head
                if verb.i in seen_verbs:
                    continue
                seen_verbs.add(verb.i)
                # Skip stative-OK verbs unless they have an explicit agent.
                # "X is bounded" → skip. "X is bounded by C" → flag.
                verb_lemma = verb.lemma_.lower()
                verb_text = verb.text.lower()
                if (verb_lemma in STATIVE_OK_VERBS or verb_text in STATIVE_OK_VERBS) \
                        and not has_agent(verb):
                    continue
                # Build a small snippet around the verb
                start = max(0, verb.i - 2)
                end = min(len(doc), verb.i + 3)
                snippet = " ".join(t.text for t in doc[start:end])
                findings.append(
                    {
                        "file": file_label,
                        "line": i,
                        "kind": "passive",
                        "verb": verb.text,
                        "aux": tok.text,
                        "snippet": snippet[:80].strip(),
                    }
                )
    return findings


def main():
    if len(sys.argv) < 2:
        print("usage: lint-passive.py <tex-file>", file=sys.stderr)
        sys.exit(2)
    path = sys.argv[1]
    with open(path, "r", encoding="utf-8") as f:
        text = f.read()
    findings = lint_text(text, path)
    json.dump(findings, sys.stdout)


if __name__ == "__main__":
    main()

File: server/lib/lint-passive.mjs
// Passive-voice lint for agent-written prose in .tex files.
//
// Wraps a Python helper (lint-passive.py) that uses spaCy dependency
// parsing — auxpass/nsubjpass relations — to detect passive constructions
// with high accuracy. Math environments and inline math are stripped before
// parsing so equations don't trip the parser.

import { spawnSync } from 'node:child_process'
import { writeFileSync, mkdtempSync, realpathSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const HELPER = join(__dirname, 'lint-passive.py')

// Run the Python helper on a single file's contents. Returns the parsed
// findings array, or [] on any error (logged).
function runHelper(filePath) {
  const result = spawnSync('python3', [HELPER, filePath], { encoding: 'utf8' })
  if (result.status !== 0) {
    console.error('[lint-passive] helper failed:', result.stderr || `exit ${result.status}`)
    return []
  }
  try {
    return JSON.parse(result.stdout || '[]')
  } catch (e) {
    console.error('[lint-passive] failed to parse helper output:', e.message)
    return []
  }
}

// Lint a full text. Writes to a temp file (the helper takes a path so it
// can preserve line numbers cleanly) and runs the Python helper.
export function lintText(text, file = '<text>') {
  const dir = mkdtempSync(join(tmpdir(), 'lint-passive-'))
  const tmpFile = join(dir, 'input.tex')
  writeFileSync(tmpFile, text)
  const raw = runHelper(tmpFile)
  // Re-label findings with the caller-provided file label, since the helper
  // gets the temp path.
  return raw.map((f) => ({ ...f, file, pattern: f.kind || 'passive' }))
}

// Parse a unified diff and return only added lines per file with their
// new-file line numbers. Returns Map<file, Set<lineNumber>>.
function parseDiffAddedLines(diffText) {
  const out = new Map()
  let curFile = null
  let curLine = 0
  let added
  for (const l of diffText.split('\n')) {
    const fileMatch = l.match(/^\+\+\+ (?:b\/)?(.+)$/)
    if (fileMatch) {
      curFile = fileMatch[1]
      added = new Set()
      out.set(curFile, added)
      continue
    }
    if (!added) continue
    const hunk = l.match(/^@@ [^+]*\+(\d+)(?:,\d+)? @@/)
    if (hunk) {
      curLine = parseInt(hunk[1], 10)
      continue
    }
    if (l.startsWith('+') && !l.startsWith('+++')) {
      added.add(curLine)
      curLine++
    } else if (l.startsWith('-') && !l.startsWith('---')) {
      // doesn't advance new-file line
    } else if (!l.startsWith('\\')) {
      curLine++
    }
  }
  return out
}

// Lint a unified diff against the post-state file contents.
// `diffText` — output of `git diff` or similar
// `readFile(path)` — function returning post-state file contents (or null)
// Returns flat array of lint results with file/line/pattern/snippet.
export function lintDiff(diffText, readFile) {
  const addedByFile = parseDiffAddedLines(diffText)
  const results = []
  for (const [file, addedLines] of addedByFile.entries()) {
    if (!file.endsWith('.tex')) continue
    let content
    try { content = readFile(file) } catch { content = null }
    if (!content) continue
    const allFindings = lintText(content, file)
    for (const finding of allFindings) {
      if (addedLines.has(finding.line)) results.push(finding)
    }
  }
  return results
}

// Standalone CLI — pipe diff to stdin, set TLDA_SRCDIR to post-state source dir
if (process.argv[1] && realpathSync(process.argv[1]) === fileURLToPath(import.meta.url)) {
  const { readFileSync } = await import('fs')
  const { join } = await import('path')
  const srcDir = process.env.TLDA_SRCDIR || ''
  const diffText = readFileSync(0, 'utf8')
  const findings = lintDiff(diffText, (file) => {
    if (!srcDir) return null
    try { return readFileSync(join(srcDir, file), 'utf8') } catch { return null }
  })
  if (findings.length > 0) {
    const grouped = new Map()
    for (const f of findings) {
      if (!grouped.has(f.file)) grouped.set(f.file, [])
      grouped.get(f.file).push(f)
    }
    const lines = [`🟡 **Passive-voice lint** — ${findings.length} passive construction(s) flagged in this build:`]
    for (const [file, items] of grouped) {
      const detail = items.map((i) => `L${i.line} (${i.pattern}): ${i.snippet}`).join('; ')
      lines.push(`- \`${file}\`: ${detail}`)
    }
    lines.push('Passive voice in math prose makes the reader work harder. Prefer active: "we bound …" / "Cauchy–Schwarz gives …" over "is bounded by …" / "is given by …".')
    process.stdout.write(lines.join('\n') + '\n')
  }
}
